### PR TITLE
Add TimeFormat and Timestamp

### PR DESCRIPTION
**Pull request by MinnDevelopment**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [ ] Internal code
- [x] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

This adds utilities to parse and format timestamps in Discord markdown.

### Example

```java
channel.sendMessage("The beginning of time: " + TimeFormat.RELATIVE.format(0L)).queue();
```

image

image

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/api/utils/TimeFormat.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.TemporalAccessor;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public enum TimeFormat
+{
+    /** Formats time as {@code 18:49} or {@code 6:49 PM} */
+    TIME_SHORT("t"),
```
Comment (duncte123) [inline src/main/java/net/dv8tion/jda/api/utils/TimeFormat.java:31]: Couldn't this be a char?

Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/api/utils/TimeFormat.java:31]: That would make the enum less robust for no real benefit

File: src/main/java/net/dv8tion/jda/api/utils/Timestamp.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    public Timestamp minus(long millis)
+    {
+        return new Timestamp(format, timestamp - millis);
+    }
+
+    @Nonnull
+    public Timestamp minus(@Nonnull Duration duration)
+    {
+        Checks.notNull(duration, "Duration");
+        return plus(duration.toMillis());
```
Comment (anweisen) [inline src/main/java/net/dv8tion/jda/api/utils/Timestamp.java:77]:
```suggestion
        return minus(duration.toMillis());
```

File: src/main/java/net/dv8tion/jda/api/utils/TimeFormat.java [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    /** Formats time as {@code 18:49:26} or {@code 6:49:26 PM} */
+    TIME_LONG("T"),
+    /** Formats date as {@code 16/06/2021} or {@code 06/16/2021} */
+    DATE_SHORT("d"),
+    /** Formats date as {@code 16 June 2021} */
+    DATE_LONG("D"),
+    /** Formats date and time as {@code 16 June 2021 18:49} or {@code 16 June 2021 6:49 PM} */
```
Comment (Chew) [inline src/main/java/net/dv8tion/jda/api/utils/TimeFormat.java:51]:
```suggestion
    /** Formats date and time as {@code 16 June 2021 18:49} or {@code June 16, 2021 6:49 PM} */
```
This rendered as the default value for me with English US

`<t:1624924743:f> -> June 28, 2021 6:59 PM`

File: src/main/java/net/dv8tion/jda/api/utils/TimeFormat.java [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    DATE_TIME_SHORT("f"),
+    /** Formats date and time as {@code Wednesday, 16 June 2021 18:49} or {@code Wednesday, 16 June 2021 6:49 PM} */
```
Comment (Chew) [inline src/main/java/net/dv8tion/jda/api/utils/TimeFormat.java:53]:
```suggestion
    /** Formats date and time as {@code Wednesday, 16 June 2021 18:49} or {@code Wednesday, June 16, 2021 6:49 PM} */
```

Same as above,
`<t:1624924743:F> -> Monday, June 28, 2021 6:59 PM`

File: src/main/java/net/dv8tion/jda/api/utils/TimeFormat.java [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        for (TimeFormat format : values())
+        {
+            if (format.style.equals(style))
+                return format;
+        }
+        return DEFAULT;
+    }
+
+    /**
+     * Parses the provided markdown into an {@link Timestamp} instance.
```
Comment (Chew) [inline src/main/java/net/dv8tion/jda/api/utils/TimeFormat.java:140]:
```suggestion
     * Parses the provided markdown into a {@link Timestamp} instance.
```

File: src/main/java/net/dv8tion/jda/api/utils/Timestamp.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+import javax.annotation.Nonnull;
+import java.time.Duration;
+import java.time.Instant;
+
+/**
+ * Utility class representing Discord Markdown timestamps.
+ * <br>This class implements {@link #toString()} such that it can be directly included in message content.
+ *
+ * <p>These timestamps are rendered by the individual receiving discord client in a local timezone and language format.
```
Comment (Chew) [inline src/main/java/net/dv8tion/jda/api/utils/Timestamp.java:29]:
```suggestion
 * <p>These timestamps are rendered by the individual receiving Discord client in a local timezone and language format.
```

**Review by DV8FromTheWorld — APPROVED** [review]

**Pull request merged**